File: libs/cua-driver/examples/jev-use/python/choose_action.py
from __future__ import annotations

import json
import math
import re
import sys
from typing import Any, Mapping

from jev_adapter import ProviderChoice, choose_bounded_with_typesafe

REQUEST_SCHEMA = "cua.jev_choice_request_v1"
RESPONSE_SCHEMA = "cua.jev_choice_v1"
MAX_INPUT_BYTES = 65_536
MAX_CANDIDATES = 32
MAX_REGIONS = 100
MAX_HISTORY = 16
ID_PATTERN = re.compile(r"[A-Za-z0-9][A-Za-z0-9._:-]{0,63}\Z")


def _record(value: Any, message: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(message)
    return value


def _bounded_string(value: Any, name: str, limit: int) -> str:
    if not isinstance(value, str) or not value.strip() or len(value) > limit:
        raise ValueError(f"{name} must be a nonempty string of at most {limit} characters")
    return value


def _identifier(value: Any, name: str) -> str:
    value = _bounded_string(value, name, 64)
    if not ID_PATTERN.fullmatch(value):
        raise ValueError(f"{name} contains unsupported characters")
    return value


def validate_request(value: Any) -> dict[str, Any]:
    root = _record(value, "request must be a JSON object")
    expected_keys = {"schema", "goal", "capture_id", "regions", "history", "candidates"}
    if set(root) != expected_keys or root.get("schema") != REQUEST_SCHEMA:
        raise ValueError(f"request must match {REQUEST_SCHEMA}")

    goal = _bounded_string(root["goal"], "goal", 4_000)
    capture_id = _bounded_string(root["capture_id"], "capture_id", 256)

    raw_regions = root["regions"]
    if not isinstance(raw_regions, list) or len(raw_regions) > MAX_REGIONS:
        raise ValueError(f"regions must be an array of at most {MAX_REGIONS} items")
    regions: list[dict[str, Any]] = []
    region_ids: set[str] = set()
    for raw_value in raw_regions:
        raw = _record(raw_value, "region must be an object")
        allowed = {"id", "kind", "bounds", "text", "label", "confidence", "interactive"}
        required = {"id", "kind", "bounds", "confidence", "interactive"}
        if not set(raw).issubset(allowed) or not required.issubset(raw):
            raise ValueError("region has unsupported or missing fields")
        region_id = _bounded_string(raw["id"], "region id", 256)
        if region_id in region_ids:
            raise ValueError("region IDs must be unique")
        region_ids.add(region_id)
        kind = raw["kind"]
        if kind not in {"text", "icon"}:
            raise ValueError("region kind must be text or icon")
        bounds = _record(raw["bounds"], "region bounds must be an object")
        if set(bounds) != {"x", "y", "width", "height"}:
            raise ValueError("region bounds have unsupported or missing fields")
        for key in ("x", "y", "width", "height"):
            number = bounds[key]
            minimum = 0 if key in {"x", "y"} else 1
            if not isinstance(number, int) or isinstance(number, bool) or number < minimum:
                raise ValueError("region bounds must contain valid integers")
        text = raw.get("text")
        label = raw.get("label")
        if text is not None:
            text = _bounded_string(text, "region text", 1_000)
        if label is not None:
            label = _bounded_string(label, "region label", 1_000)
        if (kind == "text" and text is None) or (kind == "icon" and label is None):
            raise ValueError("region is missing content required by its kind")
        confidence = raw["confidence"]
        if (
            not isinstance(confidence, (int, float))
            or isinstance(confidence, bool)
            or not math.isfinite(float(confidence))
            or not 0 <= float(confidence) <= 1
        ):
            raise ValueError("region confidence must be between zero and one")
        if not isinstance(raw["interactive"], bool):
            raise ValueError("region interactive must be boolean")
        regions.append(
            {
                "id": region_id,
                "kind": kind,
                "bounds": dict(bounds),
                "text": text,
                "label": label,
                "confidence": float(confidence),
                "interactive": raw["interactive"],
            }
        )

    history = root["history"]
    if not isinstance(history, list) or len(history) > MAX_HISTORY:
        raise ValueError(f"history must be an array of at most {MAX_HISTORY} items")
    compact_history: list[dict[str, str]] = []
    for raw_value in history:
        raw = _record(raw_value, "history item must be an object")
        if not raw or not set(raw).issubset({"selected_id", "outcome"}):
            raise ValueError("history contains a forbidden field")
        item: dict[str, str] = {}
        if "selected_id" in raw:
            item["selected_id"] = _identifier(raw["selected_id"], "history selected_id")
        if "outcome" in raw:
            item["outcome"] = _bounded_string(raw["outcome"], "history outcome", 128)
        compact_history.append(item)

    raw_candidates = root["candidates"]
    if (
        not isinstance(raw_candidates, list)
        or not 2 <= len(raw_candidates) <= MAX_CANDIDATES
    ):
        raise ValueError(f"candidates must contain between 2 and {MAX_CANDIDATES} items")
    candidates: list[dict[str, str]] = []
    candidate_ids: set[str] = set()
    for raw_value in raw_candidates:
        raw = _record(raw_value, "candidate must be an object")
        if set(raw) != {"id", "description"}:
            raise ValueError("candidate may contain only id and description")
        candidate_id = _identifier(raw["id"], "candidate id")
        if candidate_id in candidate_ids:
            raise ValueError("candidate IDs must be unique")
        candidate_ids.add(candidate_id)
        candidates.append(
            {
                "id": candidate_id,
                "description": _bounded_string(raw["description"], "description", 1_000),
            }
        )
    if not {"reobserve", "abstain"}.issubset(candidate_ids):
        raise ValueError("candidates must include reobserve and abstain")

    return {
        "goal": goal,
        "capture_id": capture_id,
        "regions": regions,
        "history": compact_history,
        "candidates": candidates,
    }


def choose_request(
    request: Mapping[str, Any], *, client: Any | None = None, mock: bool = False
) -> dict[str, Any]:
    validated = validate_request(dict(request))
    criteria = {item["id"]: item["description"] for item in validated["candidates"]}
    if mock:
        selected_id = next(
            (
                candidate_id
                for candidate_id in criteria
                if candidate_id not in {"reobserve", "abstain"}
            ),
            "reobserve",
        )
        choice = ProviderChoice(
            selected_id=selected_id,
            confidence=1.0,
            probabilities={
                candidate_id: float(candidate_id == selected_id)
                for candidate_id in criteria
            },
            model="mock",
        )
    else:
        if client is None:
            from typesafe_sdk import TypeSafeClient

            with TypeSafeClient() as live_client:
                choice = choose_bounded_with_typesafe(
                    live_client,
                    goal=validated["goal"],
                    observation={
                        "capture_id": validated["capture_id"],
                        "regions": validated["regions"],
                        "history": validated["history"],
                    },
                    criteria=criteria,
                )
        else:
            choice = choose_bounded_with_typesafe(
                client,
                goal=validated["goal"],
                observation={
                    "capture_id": validated["capture_id"],
                    "regions": validated["regions"],
                    "history": validated["history"],
                },
                criteria=criteria,
            )
    return {
        "schema": RESPONSE_SCHEMA,
        "selected_id": choice.selected_id,
        "model": choice.model,
        "confidence": choice.confidence,
        "probabilities": choice.probabilities,
    }


def main() -> None:
    if sys.argv[1:] not in ([], ["--mock"]):
        raise SystemExit("usage: choose_action.py [--mock]")
    raw = sys.stdin.buffer.read(MAX_INPUT_BYTES + 1)
    if len(raw) > MAX_INPUT_BYTES:
        raise SystemExit("request exceeds input limit")
    try:
        request = json.loads(raw)
        validate_request(request)
    except (UnicodeDecodeError, json.JSONDecodeError, ValueError, TypeError, KeyError) as error:
        raise SystemExit(f"invalid request: {error}") from None
    try:
        response = choose_request(request, mock=sys.argv[1:] == ["--mock"])
    except Exception:
        raise SystemExit("provider request failed") from None
    sys.stdout.write(json.dumps(response, separators=(",", ":")) + "\n")


if __name__ == "__main__":
    main()

File: libs/cua-driver/examples/jev-use/typescript/choose_action.ts
import { readFileSync } from 'node:fs';
import { pathToFileURL } from 'node:url';

import { TypeSafeClient } from '@typesafe-ai/sdk';

import { chooseBoundedWithTypeSafe, type ProviderChoice } from './jev_adapter.js';

export const REQUEST_SCHEMA = 'cua.jev_choice_request_v1';
export const RESPONSE_SCHEMA = 'cua.jev_choice_v1';
const MAX_INPUT_BYTES = 65_536;
const MAX_CANDIDATES = 32;
const MAX_REGIONS = 100;
const MAX_HISTORY = 16;
const ID_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,63}$/;

type JsonRecord = Record<string, unknown>;
type ValidatedRequest = {
  goal: string;
  capture_id: string;
  regions: JsonRecord[];
  history: unknown[];
  candidates: { id: string; description: string }[];
};

function record(value: unknown, message: string): JsonRecord {
  if (!value || typeof value !== 'object' || Array.isArray(value)) throw new Error(message);
  return value as JsonRecord;
}

function boundedString(value: unknown, name: string, limit: number): string {
  if (typeof value !== 'string' || !value.trim() || value.length > limit) {
    throw new Error(`${name} must be a nonempty string of at most ${limit} characters`);
  }
  return value;
}

function identifier(value: unknown, name: string): string {
  const result = boundedString(value, name, 64);
  if (!ID_PATTERN.test(result)) throw new Error(`${name} contains unsupported characters`);
  return result;
}

function exactKeys(value: JsonRecord, expected: readonly string[]): boolean {
  const actual = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    actual.length === sortedExpected.length &&
    actual.every((key, index) => key === sortedExpected[index])
  );
}

export function validateRequest(value: unknown): ValidatedRequest {
  const root = record(value, 'request must be a JSON object');
  const rootKeys = ['candidates', 'capture_id', 'goal', 'history', 'regions', 'schema'];
  if (!exactKeys(root, rootKeys) || root.schema !== REQUEST_SCHEMA) {
    throw new Error(`request must match ${REQUEST_SCHEMA}`);
  }
  const goal = boundedString(root.goal, 'goal', 4_000);
  const captureId = boundedString(root.capture_id, 'capture_id', 256);

  if (!Array.isArray(root.regions) || root.regions.length > MAX_REGIONS) {
    throw new Error(`regions must be an array of at most ${MAX_REGIONS} items`);
  }
  const regionIds = new Set<string>();
  const regions = root.regions.map((item) => {
    const raw = record(item, 'region must be an object');
    const allowed = new Set(['id', 'kind', 'bounds', 'text', 'label', 'confidence', 'interactive']);
    if (
      Object.keys(raw).some((key) => !allowed.has(key)) ||
      !['id', 'kind', 'bounds', 'confidence', 'interactive'].every((key) => key in raw)
    ) {
      throw new Error('region has unsupported or missing fields');
    }
    const id = boundedString(raw.id, 'region id', 256);
    if (regionIds.has(id)) throw new Error('region IDs must be unique');
    regionIds.add(id);
    if (raw.kind !== 'text' && raw.kind !== 'icon') {
      throw new Error('region kind must be text or icon');
    }
    const bounds = record(raw.bounds, 'region bounds must be an object');
    if (!exactKeys(bounds, ['height', 'width', 'x', 'y'])) {
      throw new Error('region bounds have unsupported or missing fields');
    }
    for (const key of ['x', 'y', 'width', 'height'] as const) {
      const number = bounds[key];
      const minimum = key === 'x' || key === 'y' ? 0 : 1;
      if (!Number.isInteger(number) || Number(number) < minimum) {
        throw new Error('region bounds must contain valid integers');
      }
    }
    const text =
      raw.text === undefined || raw.text === null
        ? null
        : boundedString(raw.text, 'region text', 1_000);
    const label =
      raw.label === undefined || raw.label === null
        ? null
        : boundedString(raw.label, 'region label', 1_000);
    if ((raw.kind === 'text' && text === null) || (raw.kind === 'icon' && label === null)) {
      throw new Error('region is missing content required by its kind');
    }
    if (
      typeof raw.confidence !== 'number' ||
      !Number.isFinite(raw.confidence) ||
      raw.confidence < 0 ||
      raw.confidence > 1
    ) {
      throw new Error('region confidence must be between zero and one');
    }
    if (typeof raw.interactive !== 'boolean') {
      throw new Error('region interactive must be boolean');
    }
    return {
      id,
      kind: raw.kind,
      bounds: { ...bounds },
      text,
      label,
      confidence: raw.confidence,
      interactive: raw.interactive,
    };
  });

  if (!Array.isArray(root.history) || root.history.length > MAX_HISTORY) {
    throw new Error(`history must be an array of at most ${MAX_HISTORY} items`);
  }
  const history = root.history.map((item) => {
    const raw = record(item, 'history item must be an object');
    if (
      Object.keys(raw).length === 0 ||
      Object.keys(raw).some((key) => key !== 'selected_id' && key !== 'outcome')
    ) {
      throw new Error('history contains a forbidden field');
    }
    return {
      ...('selected_id' in raw
        ? { selected_id: identifier(raw.selected_id, 'history selected_id') }
        : {}),
      ...('outcome' in raw
        ? { outcome: boundedString(raw.outcome, 'history outcome', 128) }
        : {}),
    };
  });

  if (
    !Array.isArray(root.candidates) ||
    root.candidates.length < 2 ||
    root.candidates.length > MAX_CANDIDATES
  ) {
    throw new Error(`candidates must contain between 2 and ${MAX_CANDIDATES} items`);
  }
  const candidateIds = new Set<string>();
  const candidates = root.candidates.map((item) => {
    const raw = record(item, 'candidate must be an object');
    if (!exactKeys(raw, ['description', 'id'])) {
      throw new Error('candidate may contain only id and description');
    }
    const id = identifier(raw.id, 'candidate id');
    if (candidateIds.has(id)) throw new Error('candidate IDs must be unique');
    candidateIds.add(id);
    return { id, description: boundedString(raw.description, 'description', 1_000) };
  });
  if (!candidateIds.has('reobserve') || !candidateIds.has('abstain')) {
    throw new Error('candidates must include reobserve and abstain');
  }
  return { goal, capture_id: captureId, regions, history, candidates };
}

export async function chooseRequest(
  request: unknown,
  options: { client?: Pick<TypeSafeClient, 'systemOne'>; mock?: boolean } = {}
) {
  const validated = validateRequest(request);
  const criteria = Object.fromEntries(
    validated.candidates.map(({ id, description }) => [id, description])
  );
  let result: ProviderChoice;
  if (options.mock) {
    const selectedId =
      Object.keys(criteria).find((id) => id !== 'reobserve' && id !== 'abstain') ??
      'reobserve';
    result = {
      selectedId,
      model: 'mock',
      confidence: 1,
      probabilities: Object.fromEntries(
        Object.keys(criteria).map((id) => [id, Number(id === selectedId)])
      ),
    };
  } else {
    result = await chooseBoundedWithTypeSafe(
      options.client ?? new TypeSafeClient(),
      validated.goal,
      {
        capture_id: validated.capture_id,
        regions: validated.regions,
        history: validated.history,
      },
      criteria
    );
  }
  return {
    schema: RESPONSE_SCHEMA,
    selected_id: result.selectedId,
    model: result.model ?? null,
    confidence: result.confidence,
    probabilities: result.probabilities,
  };
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length > 1 || (args.length === 1 && args[0] !== '--mock')) {
    throw new Error('usage: choose_action.ts [--mock]');
  }
  const raw = readFileSync(0, 'utf8');
  if (Buffer.byteLength(raw, 'utf8') > MAX_INPUT_BYTES) {
    throw new Error('request exceeds input limit');
  }
  const response = await chooseRequest(JSON.parse(raw), { mock: args[0] === '--mock' });
  process.stdout.write(`${JSON.stringify(response)}\n`);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch(() => {
    process.stderr.write('chooser failed\n');
    process.exitCode = 1;
  });
}
